lib/Tools/circt-verilog-lsp-server/VerilogServerImpl/VerilogServer.cpp

```diff
@@ -358,8 +477,24 @@ llvm::SMLoc VerilogDocument::getSMLoc(slang::SourceLocation loc) {
   auto bufferIDMapIt = bufferIDMap.find(bufferID);
   if (bufferIDMapIt == bufferIDMap.end()) {
     // If not, open the source file and add it to the LLVM source manager.
-    auto path = getSlangSourceManager().getFullPath(loc.buffer());
-    auto memBuffer = llvm::MemoryBuffer::getFile(path.string());
+    const slang::SourceManager &ssm = getSlangSourceManager();
+    auto path = ssm.getFullPath(loc.buffer());
+
+    std::string name;
+    if (path.empty()) {
+      // Get filename from slang source manager - might've been loaded already!
+      llvm::StringRef fname = ssm.getFileName(loc);
+      llvm::SmallString<256> realPath;
+      if (!llvm::sys::fs::real_path(fname, realPath)) {
+        name = realPath.str().str();
+      } else {
+        name = fname.str();
+      }
+    } else {
+      name = path.string();
+    }
+
+    auto memBuffer = llvm::MemoryBuffer::getFile(name);
```

Comment:
I think we can look up the main buffer from here instead of opening a new memory buffer
```cpp
if (getFileName(uri) == name) { 
  bufferIDMap[bufferID] = 0; // Insert the id for the next call.  
  const auto *buffer = sourceMgr.getMemoryBuffer(0); // get main buffer
  return return llvm::SMLoc::getFromPointer(buffer->getBufferStart() + loc.offset());
}